### PR TITLE
[Metricbeat/Kibana/stats] `exclude_usage=true` by default

### LINKED ISSUE #22651: [Metricbeat][Kibana][stats] Stop collecting usage
**Describe the enhancement:**

As part of the efforts for decoupling Monitoring & Telemetry in Kibana (https://github.com/elastic/kibana/issues/83626), we are stopping the collection of usage data (https://github.com/elastic/kibana/issues/83521), meaning we'll only collect _metric-related_ stats from 7.11 and future versions.

This can be achieved by always sending `exclude_usage=true` when requesting the API `/api/stats?extended=true&legacy=true&exclude_usage=true` ([line where we control it](https://github.com/elastic/beats/blob/master/metricbeat/module/kibana/stats/stats.go#L166)).

**Describe a specific use case for the enhancement or feature:**

Currently, the `stats` dataset in the Metricbeat-Kibana module collects the extended usage details every 24h so the Monitoring plugin store it in the `.monitoring-kibana-*` indices. These usage details are not actually used by the Monitoring plugin itself to anything other than reporting telemetry.

With the rise of the `local` telemetry, `monitoring`-sourced telemetry doesn't need to report this data anymore, so we can disable the collection in the 2 sources of information for the Monitoring plugin:

- Legacy Kibana-driven collection (https://github.com/elastic/kibana/pull/83546)
- Metricbeat collection (this issue).

### PR DESCRIPTION
**Pull request by afharo**
## What does this PR do?

It adds changes to the Metricbeat's `kibana` module: it adds a new configuration flag `stats.exclude_usage` (default `true`) that is used in the metricset `stats` in the method `shouldCollectUsage` to decide whether it should collect usage (no matter the time).

The reason for this change is to disable the usage collection via the `/api/stats` endpoint but to still allow the collection of the usage via explicitly setting the flag to `false` if necessary.

## Why is it important?

We are working towards decoupling Monitoring and Telemetry in Kibana. At the moment, Monitoring holds in the `.monitoring-kibana-*` indices the Kibana's Usage details only to be able to report them when sending the telemetry reports. This data is already sent to the remote telemetry cluster via Kibana's `local` collectors, so there's no need for this duplication in the efforts. Thanks to these changes, Monitoring will only maintain in its indices the data they really need for its own features.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ] Did I filled the default config in the right way?
- [ ] Not a Go developer myself. Did I mess anything up? 😇 

## How to test this PR locally

Setup Metricbeat with the `kibana-xpack` module enabled and notice it never asks for the `/api/stats?extended&legacy&exclude_usage=false` request (the last query is always `true` by default).

## Related issues

- Closes #22651
- Relates https://github.com/elastic/kibana/issues/83521
- Relates https://github.com/elastic/kibana/issues/83626

## Use cases

Explained in the issue #22651

### CODE REVIEW

**Comment by andresrc** [comment]
cc @sayden for awareness given the ongoing changes

**Comment by afharo** [comment]
cc @chrisronline for awareness

**Comment by chrisronline** [comment]
@sayden @ycombinator This change relates a larger initiative we are undertaking to decouple telemetry and monitoring. For Kibana, we will no longer collect and index `usage` data, as we will rely entirely in local telemetry collection within Kibana instances (which is enabled by default). See https://github.com/elastic/kibana/issues/83626 for more information

File: metricbeat/module/kibana/mtest/testing.go [hunk 1]
```diff
@@ -27,6 +27,9 @@ func GetConfig(metricset string, host string, xpackEnabled bool) map[string]inte
 	if xpackEnabled {
 		config["xpack.enabled"] = true
 	}
+	if statsExcludeUsage == false {
```
Comment (kvch) [inline metricbeat/module/kibana/mtest/testing.go:30]:
```suggestion
	if !statsExcludeUsage {
```

File: metricbeat/module/kibana/stats/stats.go [hunk 2]
```diff
@@ -221,5 +224,6 @@ func (m *MetricSet) calculateIntervalMs() int64 {
 }
 
 func (m *MetricSet) shouldCollectUsage(now time.Time) bool {
-	return now.Sub(m.usageLastCollectedOn) > usageCollectionPeriod && now.Sub(m.usageNextCollectOn) > 0
+	return m.statsExcludeUsage == false && // If the config claims we should exclude the usage, there is no need to check for the time
```
Comment (kvch) [inline metricbeat/module/kibana/stats/stats.go:227]:
```suggestion
	return !m.statsExcludeUsage && // If the config claims we should exclude the usage, there is no need to check for the time
```

**Comment by ycombinator** [comment]
Hey @afharo, thanks for this PR.

I understand the reasons for this change and am generally ++ for it. However, I'd like to explore the possibility of not introducing a configuration setting to toggle this change. Instead, I'd like to explore if we could programmatically determine the right value for the `exclude_usage` flag based on some conditions.

With the change in this PR, the `kibana/stats` metricset will no longer be responsible for collecting usage stats from Kibana and indexing it into Stack Monitoring indices. And subsequently, Telemetry code in Kibana will no longer have access to these usage stats from the Stack Monitoring indices. So I imagine there is a parallel effort happening in Kibana to get the same usage stats into Telemetry some other way.  Which version of Kibana is this parallel effort expected to land in? Maybe we could toggle the `exclude_usage` flag in the `kibana/stats` metricset based on this version (we do this already in a couple of places: https://github.com/elastic/beats/blob/0d4a830fea46210ee264c52a977834d39493c750/metricbeat/module/kibana/kibana.go#L92-L100). Essentially what we'll be saying is: starting from version X, Kibana knows how to collect usage stats on its own, so no need for the `kibana/stats` metricset to collect it.

**Comment by afharo** [comment]
Hey @ycombinator, 

Thanks for coming back to me! These are very valid points! Our intention is to ship the parallel effort in Kibana 7.11 (almost there 😅).

I'd love to use your knowledge here to deep dive on this version-dependency: The way Telemetry will use the Monitoring data depends on the Monitoring cluster's version, but not the _monitored_ production clusters. This means that:

- If the Monitoring cluster is `7.11` and all the production clusters are running any other version, the telemetry will be fine.
- If the Production cluster is `7.11`, but the Monitoring cluster is not at least `7.11`, the reported telemetry might be incomplete.

Do you know if Metricbeat is, somehow, aware of the version of the Monitoring cluster it is sending the stats to?

**Comment by ycombinator** [comment]
Ah, thanks for clarifying the monitoring vs. monitored cluster situation here. There's no way for Metricbeat to know which monitoring cluster it's sending the data to so we can't check it's version.

Given that, I think what you have in this PR is a good approach for now - we default to the current behavior but provide a flag to turn it off in a future release. One question then becomes: do we need documentation to tell users to explicitly set the setting to `false` if they are running `7.11` or greater in their monitoring cluster?

Alternatively, we could just turn it off completely in this PR - i.e. no setting to toggle it - but only merge it into `master`, i.e. we won't backport it to 7.x. This will mean that starting 8.0, the `kibana/stats` metricset in Metricbeat will never collect usage and versions before that (`7.x`) will always collect usage (which is the current behavior).

**Comment by afharo** [comment]
> Given that, I think what you have in this PR is a good approach for now - we default to the current behavior but provide a flag to turn it off in a future release. 

In this PR I actually did the opposite: disabled by default, but a flag to turn it back on if needed. But I'm happy to flip the default value if you think it's safer.

> One question then becomes: do we need documentation to tell users to explicitly set the setting to `false` if they are running `7.11` or greater in their monitoring cluster?

Given that we are doing the opposite change in this PR (for now), I think we should tell our users to explicitly set `false` if their monitoring cluster is `7.10` or earlier _to maintain backwards compatiblity_.

> Alternatively, we could just turn it off completely in this PR - i.e. no setting to toggle it - but only merge it into `master`, i.e. we won't backport it to 7.x. This will mean that starting 8.0, the `kibana/stats` metricset in Metricbeat will never collect usage and versions before that (`7.x`) will always collect usage (which is the current behavior).

I think, ultimately, we'll complete this breaking change and remove the option flag entirely. Whether to keep collecting it until `8.0` or not, I don't have a strong opinion. I'd like to know @chrisronline's and @kobelb's opinions on that.

**Comment by ycombinator** [comment]
So, maybe the naming of the setting is confusing me :) 

> Given that we are doing the opposite change in this PR (for now), I think we should tell our users to explicitly set `false` if their monitoring cluster is `7.10` or earlier to maintain backwards compatiblity.

I think it's a bit odd to ask users to take some action in a minor version to maintain backwards compatibility. I think backwards compatibility should be expected without any extra actions by the user between minor version upgrades.

So I think the default value of the setting should result in the behavior we have today (which is to collect usage stats). That way, users of Metricbeat 7.11 (assuming this PR makes it to that release) should not need to do anything extra right after they upgrade just to maintain backwards compatibility. Then, we also add to documentation saying: if your monitoring cluster is >= 7.11, please set X setting to Y value for so and so reason. And from a user perspective there is a compelling reason: usage stats docs can be quite large so not collecting them via Metricbeat saves memory + size on the wire.

Then in `8.0` / `master` we can simply remove the logic from the `kibana/stats` metricset to collect usage stats as well as the setting to allow users to toggle this collection.

**Comment by afharo** [comment]
@ycombinator sounds like a very sensible plan to me! Thank you! I'll apply those changes to this PR.

**Comment by kobelb** [comment]
It's unclear to me why changing the behavior of Metricbeat to exclude the usage would be a breaking change from the end-users perspective. The output of Kibana's stats endpoint is indexed into the `.monitoring-kibana-*` and end-users shouldn't be querying these documents directly. Changing this behavior should only affect the telemetry data that is sent to Elastic itself, and we're tolerant of this change in a minor version.

**Comment by ycombinator** [comment]
Fair point @kobelb. I've been considering both the end-user and internal (Elastic) users both as users of this module. Asking the `kibana/stats` Metricbeat module to stop collecting usage telemetry is definitely not a breaking change for _end-users_. And if _internal users_ are tolerant to not having usage telemetry in case an end-user runs Metricbeat 7.11+ against a monitoring cluster that's < 7.11 (as @afharo explained in https://github.com/elastic/beats/pull/22654#issuecomment-732119213), then I'm ++ removing usage telemetry collection in `kibana/stats` in 7.11.

In fact, I think we could simplify this PR even more then. Is there any reason we couldn't simply stop collecting usage data starting Metricbeat 7.11?

**Comment by afharo** [comment]
> And if _internal users_ are tolerant to not having usage telemetry in case an end-user runs Metricbeat 7.11+ against a monitoring cluster that's < 7.11 (as @afharo explained in [#22654 (comment)](https://github.com/elastic/beats/pull/22654#issuecomment-732119213)), then I'm ++ removing usage telemetry collection in `kibana/stats` in 7.11.

AFAIK, we should be OK in that scenario. The 7.11+ (in fact, 7.2+) production Kibana instance will be reporting their `local`ly collected telemetry anyway (that's why we are after stopping Monitoring from doing so).

> In fact, I think we could simplify this PR even more then. Is there any reason we couldn't simply stop collecting usage data starting Metricbeat 7.11?

I was a bit wary of simply deleting the piece of code. But if we are all happy to do so, tomorrow I'll create a new PR with the removal and we can choose which one we like best :)

Thank you both @kobelb and @ycombinator for your thoughts on this!

**Comment by afharo** [comment]
@ycombinator I've created https://github.com/elastic/beats/pull/22732 superseding this one. It fully removes the `exclude_usage=false` logic

**Comment by afharo** [comment]
Closing this PR in favour of #22732

**Pull request closed without merging**